src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs

```diff
@@ -136,7 +142,46 @@ public async IAsyncEnumerable<FeatureDefinition> GetAllFeatureDefinitionsAsync()
 
                 //
                 // Underlying IConfigurationSection data is dynamic so latest feature definitions are returned
-                yield return  _definitions.GetOrAdd(featureName, (_) => ReadFeatureDefinition(featureSection));
+                yield return _definitions.GetOrAdd(featureName, (_) => ReadFeatureDefinition(featureSection));
+            }
+        }
+
+        private void EnsureInit()
+        {
+            if (_initialized == 0)
+            {
+                IConfiguration MicrosoftFeatureManagementConfigurationSection = _configuration
+                    .GetChildren()
+                    .FirstOrDefault(section =>
+                        string.Equals(
+                            section.Key,
+                            MicrosoftFeatureManagementFields.FeatureManagementSectionName,
+                            StringComparison.OrdinalIgnoreCase));
+
+                bool hasMicrosoftFeatureManagementSchema = MicrosoftFeatureManagementConfigurationSection != null;
+
+                if (MicrosoftFeatureManagementConfigurationSection == null & RootConfigurationFallbackEnabled)
+                {
+                    IConfiguration featureFlagsSection = _configuration
+                        .GetChildren()
+                        .FirstOrDefault(section =>
+                            string.Equals(
+                                section.Key,
+                                MicrosoftFeatureManagementFields.FeatureFlagsSectionName,
+                                StringComparison.OrdinalIgnoreCase));
+
+                    hasMicrosoftFeatureManagementSchema = featureFlagsSection != null;
+                }
+
+                lock (_lock)
+                {
+                    if (Interlocked.Read(ref _initialized) == 0)
+                    {
+                        _microsoftFeatureManagementSchemaEnabled = hasMicrosoftFeatureManagementSchema;
+
+                        Interlocked.Exchange(ref _initialized, 1);
+                    }
```

Comment:
```suggestion
                   if (Interlocked.Exchange(ref _initialized, 1) == 0)
                    {
                        _microsoftFeatureManagementSchemaEnabled = hasMicrosoftFeatureManagementSchema;
                    }
```

I think this achieves the same logic, and does the read/exchange at the same time

Reply:
I think you might also not need the lock in this case

Reply:
I see, I don't fully understand from the comment but I didn't know it was already agreed on. I'm fine to leave it as is.

Reply:
@amerjusupovic the difference between your suggestion and what's checked in is that your suggestion can update `_microsoftFeatureManagementSchemaEnabled` even if it's already been set before.

However, there is a valid suggestion similar to what you have which is

```
                   if (Interlocked.CompareExchange(ref _initialized, 1, 0) == 0)
                    {
                        _microsoftFeatureManagementSchemaEnabled = hasMicrosoftFeatureManagementSchema;
                    }
```

It uses CompareExhange to only enter the if statement if the value of `_initialized` was `0`.

Reply:
The issue is that `_initialized` should not be set to `1` until `_microsoftFeatureManagementSchemaEnabled` has been assigned to.